File: interfaces/scenario_templates.py
"""Helpers for loading shared scenario templates across Python interfaces."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, TypedDict

_ROOT = Path(__file__).resolve().parent.parent
_DEFAULT_TEMPLATES_PATH = _ROOT / 'configs' / 'scenarios' / 'templates.json'


class ScenarioConfig(TypedDict):
    """Structure of a scenario defaults block."""

    initialBalance: float
    riskPerTrade: float
    maxPositions: int
    timeframe: str


class ScenarioTemplate(TypedDict):
    """Structure of a scenario template entry."""

    id: str
    label: str
    description: str
    defaults: ScenarioConfig
    notes: list[str]


def _validate_template(payload: dict[str, Any], index: int) -> ScenarioTemplate:
    try:
        defaults = payload['defaults']
    except KeyError as exc:  # pragma: no cover - defensive guard
        raise ValueError(f'Scenario template at index {index} is missing defaults') from exc

    if not isinstance(defaults, dict):
        raise TypeError(f'Scenario template defaults at index {index} must be an object')

    required_defaults = {'initialBalance', 'riskPerTrade', 'maxPositions', 'timeframe'}
    if not required_defaults.issubset(defaults):
        missing = required_defaults.difference(defaults)
        raise ValueError(f'Scenario template defaults at index {index} missing keys: {sorted(missing)}')

    if not isinstance(defaults['initialBalance'], (int, float)):
        raise TypeError(f'initialBalance at index {index} must be numeric')
    if not isinstance(defaults['riskPerTrade'], (int, float)):
        raise TypeError(f'riskPerTrade at index {index} must be numeric')
    if not isinstance(defaults['maxPositions'], (int, float)):
        raise TypeError(f'maxPositions at index {index} must be numeric')
    if not isinstance(defaults['timeframe'], str):
        raise TypeError(f'timeframe at index {index} must be a string')

    notes = payload.get('notes', [])
    if not isinstance(notes, list) or not all(isinstance(note, str) for note in notes):
        raise TypeError(f'notes at index {index} must be a list of strings')

    template: ScenarioTemplate = {
        'id': str(payload.get('id', '')),
        'label': str(payload.get('label', '')),
        'description': str(payload.get('description', '')),
        'defaults': {
            'initialBalance': float(defaults['initialBalance']),
            'riskPerTrade': float(defaults['riskPerTrade']),
            'maxPositions': int(defaults['maxPositions']),
            'timeframe': defaults['timeframe'],
        },
        'notes': list(notes),
    }

    for field in ('id', 'label', 'description'):
        if not template[field]:
            raise ValueError(f'Scenario template at index {index} missing required field: {field}')

    return template


def load_scenario_templates(path: str | Path | None = None) -> list[ScenarioTemplate]:
    """Load scenario templates from the shared JSON configuration."""

    location = Path(path) if path else _DEFAULT_TEMPLATES_PATH
    raw = location.read_text(encoding='utf-8')
    parsed = json.loads(raw)
    if not isinstance(parsed, list):
        raise TypeError('Scenario templates JSON must contain an array at the top level')

    templates = [_validate_template(entry, index) for index, entry in enumerate(parsed)]
    return templates


__all__ = ['ScenarioConfig', 'ScenarioTemplate', 'load_scenario_templates']

File: apps/web/lib/scenario-templates.ts
import 'server-only'

import fs from 'node:fs/promises'
import path from 'node:path'

import type { ScenarioConfig, ScenarioTemplate } from './scenario-templates-types'

const templatesPath = path.join(process.cwd(), 'configs', 'scenarios', 'templates.json')

function assertConfig(config: any, index: number): ScenarioConfig {
  if (!config || typeof config !== 'object') {
    throw new TypeError(`Scenario template at index ${index} is missing defaults`)
  }
  const { initialBalance, riskPerTrade, maxPositions, timeframe } = config as ScenarioConfig
  if (
    typeof initialBalance !== 'number' ||
    typeof riskPerTrade !== 'number' ||
    typeof maxPositions !== 'number' ||
    typeof timeframe !== 'string'
  ) {
    throw new TypeError(`Scenario template defaults at index ${index} have invalid types`)
  }
  return { initialBalance, riskPerTrade, maxPositions, timeframe }
}

export async function loadScenarioTemplates(): Promise<ScenarioTemplate[]> {
  const raw = await fs.readFile(templatesPath, 'utf-8')
  const parsed = JSON.parse(raw)
  if (!Array.isArray(parsed)) {
    throw new TypeError('Scenario templates JSON must be an array')
  }
  return parsed.map((item, index) => {
    if (!item || typeof item !== 'object') {
      throw new TypeError(`Scenario template at index ${index} must be an object`)
    }
    const { id, label, description, defaults, notes } = item as Record<string, unknown>
    if (
      typeof id !== 'string' ||
      typeof label !== 'string' ||
      typeof description !== 'string' ||
      !Array.isArray(notes) ||
      !notes.every((note) => typeof note === 'string')
    ) {
      throw new TypeError(`Scenario template at index ${index} has invalid metadata`)
    }
    return {
      id,
      label,
      description,
      defaults: assertConfig(defaults, index),
      notes,
    }
  })
}
